### PR TITLE
ISPN-5758 Clients should find out if compatibility is enabled

### PR DESCRIPTION
**Pull request by galderz**
https://issues.jboss.org/browse/ISPN-5758

* Hot Rod servers now provide different status identifiers when
  compatibility is enabled.
* A new version of the Hot Rod protocol, version 2.4, has been created
  to support the new status ids.
* Clients need to know whether compatibility is enabled or not to be
  able to apply correct unmarshalling when keys and/or values are
  returned back as part of the response.
* Version handling in both the client and server has been centralised in
  order to make it easier to do a version upgrade.

### CODE REVIEW

**Comment by gustavonalle** [comment]
There are test failures

**Comment by galderz** [comment]
@gustavonalle Tests should be fixed, marshaller was getting mixed between different tests.

**Comment by gustavonalle** [comment]
Looking

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/protocol/Codec20.java [hunk 1]
```diff
@@ -45,6 +45,13 @@
    final boolean trace = getLog().isTraceEnabled();
 
    @Override
+   public <T> T readUnmarshallByteArray(Transport transport, short status) {
+      byte[] bytes = transport.readArray();
+      Marshaller marshaller = transport.getTransportFactory().getMarshaller();
+      return MarshallerUtil.bytes2obj(marshaller, bytes, status);
+   }
```
Comment (gustavonalle) [inline client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/protocol/Codec20.java]: This method is duplicated on Codec10, could you extract it to a common place?

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/iteration/RemoteCloseableIterator.java [hunk 2]
```diff
@@ -76,30 +75,30 @@ public boolean hasNext() {
...
 
             if (entries.length == 0) {
                endOfIteration = true;
                break;
             }
-            for (Entry<byte[], byte[]> entry : entries) {
-               if (segmentKeyTracker.track(entry.getKey())) {
-                  nextElements.add(new SimpleEntry<>(unmarshall(entry.getKey()), unmarshall(entry.getValue())));
+            for (Entry<K, V> entry : entries) {
+               if (segmentKeyTracker.track(obj2bytes(marshaller, entry.getKey(), true, 100, 100))) {
```
Comment (gustavonalle) [inline client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/iteration/RemoteCloseableIterator.java]: Can this extra marshalling be avoided?
On master, it takes raw entries from ```iterationNextOperation```, uses the ```segmentKeyTracker``` to track raw keys, and finally unmarshalls K,V to serve to the user. Now it is always unmarshalling K and V plus marshalling again the key. :/

Comment (gustavonalle) [inline client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/iteration/RemoteCloseableIterator.java]: @galderz, also, could you add a snippet to the .adoc about what changed in the HotRod protocol 2.4?

**Comment by galderz** [comment]
@gustavonalle Fixed all comments

**Comment by gustavonalle** [comment]
pulling

**Comment by gustavonalle** [comment]
Integrated, thanks @galderz !

**Pull request merged**